File: CSharpCodeBase/aiactions/returnhome.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class returnhome :ainode{
 var AI_ReturnHome = Class({__includes=AINode, init = public void (self, entity){
   AINode.init(self, entity);
   this.timeout = null;
   this.radius = this.entity.config:Get("returnHomeRadius");
 }})
 public void AI_Visit(){
   if(this.entity.isChild  &&  this.entity.relationship:GetInstance("parent") != null  ){ 
     this.status = NODE_FAILURE;
     return;
   }
   
   if(this.status == NODE_READY  ){ 
     this.status = NODE_FAILURE;
     if(this.entity.worldmap:GetDistanceTo("home") > this.radius  &&  this.timeout != null  ){ 
       this.timeout = this.timeout - GameController.deltaTime;
     }else{
       this.timeout = this.entity.config:Get("returnHomeTimeout");
     }
   }
   
   if(this.timeout < 0  &&  this.status != NODE_RUNNING  ){ 
     this.status = NODE_RUNNING;
     this.entity.mover:ResetSpeed();
     this.entity.mover:SetGoal(this.entity.worldmap:GetLocation("home"));
   }else{if this.status == NODE_RUNNING  ){ 
     if(not this.entity.mover:IsHaveGoal()  ){ 
       this.status = NODE_SUCCESS;
     }
   }
   return this.status;
 }
 return AI_ReturnHome}}

File: CSharpCodeBase/aiactions/returnhome.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

local AI_ReturnHome = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
  self.timeout = nil
  self.radius = self.entity.config:Get("returnHomeRadius")
end})

function AI_ReturnHome:Visit()
  if self.entity.isChild and self.entity.relationship:GetInstance('parent') ~= nil then
    self.status = NODE_FAILURE
    return
  end
  
  if self.status == NODE_READY then
    self.status = NODE_FAILURE
    if self.entity.worldmap:GetDistanceTo("home") > self.radius and self.timeout ~= nil then
      self.timeout = self.timeout - GameController.deltaTime
    else
      self.timeout = self.entity.config:Get("returnHomeTimeout")
    end
  end
  
  if self.timeout < 0 and self.status ~= NODE_RUNNING then
    self.status = NODE_RUNNING
    self.entity.mover:ResetSpeed()
    self.entity.mover:SetGoal(self.entity.worldmap:GetLocation('home'))
  elseif self.status == NODE_RUNNING then
    if not self.entity.mover:IsHaveGoal() then
      self.status = NODE_SUCCESS
    end
  end
  return self.status
end

return AI_ReturnHome